pkg/streamingpromql/functions/range_vectors.go

```diff
@@ -50,10 +58,271 @@ var PresentOverTime = FunctionOverRangeVector{
 	StepFunc:           presentOverTime,
 }
 
-func presentOverTime(step types.RangeVectorStepData, _ float64, fPoints *types.FPointRingBuffer, hPoints *types.HPointRingBuffer, _ EmitAnnotationFunc) (f float64, hasFloat bool, h *histogram.FloatHistogram, err error) {
+func presentOverTime(step types.RangeVectorStepData, _ float64, fPoints *types.FPointRingBuffer, hPoints *types.HPointRingBuffer, _ EmitAnnotationFunc) (float64, bool, *histogram.FloatHistogram, error) {
 	if fPoints.AnyAtOrBefore(step.RangeEnd) || hPoints.AnyAtOrBefore(step.RangeEnd) {
 		return 1, true, nil, nil
 	}
 
 	return 0, false, nil, nil
 }
+
+var MaxOverTime = FunctionOverRangeVector{
+	SeriesMetadataFunc: DropSeriesName,
+	StepFunc:           maxOverTime,
+}
+
+func maxOverTime(step types.RangeVectorStepData, _ float64, fPoints *types.FPointRingBuffer, _ *types.HPointRingBuffer, _ EmitAnnotationFunc) (float64, bool, *histogram.FloatHistogram, error) {
+	head, tail := fPoints.UnsafePoints(step.RangeEnd)
+
+	if len(head) == 0 && len(tail) == 0 {
+		return 0, false, nil, nil
+	}
+
+	var maxSoFar float64
+
+	if len(head) > 0 {
```

Comment:
(nit) possible not worth it, but we have a few functions that need to check every single point. We could create a helper function that iterates over the tail+head that we pass some function into it that operates on each point.

Reply:
Once Go 1.23 has been released, I'm hoping to use https://tip.golang.org/wiki/RangefuncExperiment to do exactly that - there are a bunch of places that would benefit from this, and it should allow us to eliminate the use of `UnsafePoints`.